jsx/App/Stories/Story/Display/TextFormatButton.jsx

```diff
@@ -0,0 +1,243 @@
+            let gloss = [];
+            let flag = true; 
+            // Find the morphemes belonging to the current word, and add them and their gloss
+            // into a list. 
+            while (flag && morphemeListIndex < morphemeList.length) {
+                // Add the "Undefined" strings just in case that some texts have glossing misaligned.
+                // For texts with aligned morphemes and glossing, "Undefined" shouldn't show up,
+                // but they do to avoid an error being thrown and as a way to let the user know 
+                // that something is off with this sentence so they should edit the generated LaTeX code. 
+                const morphemeEntry = morphemeList[morphemeListIndex] || "Undefined";;
+                const glossEntry = glossList[morphemeListIndex] || "Undefined";
+                if (morphemeEntry["start_slot"] >= wordStartSlot && morphemeEntry["end_slot"] <= wordEndSlot) {
+                    // If a morpheme item has the whole word, eg. "cundyi-'je='fa", we need to
+                    // split the current morpheme on = or -, so that each root or suffix or clitic is on its own. 
+                    const morphemeValue = morphemeEntry["value"] || "Undefined";
+                    const glossValue = glossEntry["value"] || "Undefined";
+                    // First, add a space in front of = and - so that we can split on space later and preserve both = and -
+                    const currentMorpheme = morphemeValue.replace("=", " =").replace("-", " -");
+                    const currentGloss = glossValue.replace("=", " =").replace("-", " -");;          
+                    const currentMorphemeSplit = currentMorpheme.split(" ");
+                    const currentGlossSplit = currentGloss.split(" ");
+
+                    for (const e of currentMorphemeSplit) {
+                        if (e !== "") {
+                            morphemes.push(e);
+                        }
+                    }
+                    for (const e of currentGlossSplit) {
+                        if (e !== "") {
+                            gloss.push(e);
+                        }
+                    }
+                    morphemeListIndex += 1;
+                } else {   
+                    flag = false;
+                } 
+            }
+            word2Morpheme[wordListCounter][word] = morphemes;
+            word2Gloss[wordListCounter][word] = gloss;
+            wordListCounter += 1;
+        }
+
+        return {
+            "morphemes" : word2Morpheme,
+            "gloss" : word2Gloss
+        };
+    }
+
+    /* Retrieves the English translation for the sentence. */
+    function getSentenceTranslation(dependents) {
+        // Find the tier that corresponds to the English translation. 
+        for (let i = 0; i < dependents.length; i+=1) {
+            const tierName = dependents[i]["tier"].toLowerCase();
+            if (tierName == "frase en inglés" || tierName == "english") {
```

Comment:
The glossing language isn't always English. (In our corpora, it's always English and/or Spanish, but LingView can also be used with other languages.) Even when it's English, the gloss tier can be named anything. If the file comes from ELAN, we use whatever tier names were in the original document.